File: backend/utils/oauth_utils.py
import logging
from datetime import datetime
from extensions import supabase

def validate_state(state):
    """
    Validates the incoming OAuth state against chatgpt_oauth_states (CSRF protection).
    Ensures that the state exists and hasn't expired.
    """
    response = supabase.table("chatgpt_oauth_states").select("*").eq("state", state).execute()
    if not response.data:
        logging.error(f"❌ State not found in database: {state}")
        raise ValueError("Invalid or expired state parameter.")

    stored_state = response.data[0]
    expiry = datetime.fromisoformat(stored_state["expiry"])
    if datetime.utcnow() > expiry:
        logging.error(f"State expired. Generated: {stored_state['expiry']} Current: {datetime.utcnow()}")
        raise ValueError("State token expired.")
    
    return stored_state # ✅ Returns stored session details if valid

File: backend-node/src/utils/oauthUtils.js
const supabase = require("../config/supabaseClient");
const logger = require("./loggingUtils");

logger.info(`✅ Validating OAuth state: ${state}`);

/**
 * Validates the incoming OAuth state against `chatgpt_oauth_states`.
 * Ensures the state exists and hasn't expired.
 */
const validateState = async (state) => {
  const { data, error } = await supabase
    .from("chatgpt_oauth_states")
    .select("*")
    .eq("state", state)
    .single();

  if (error || !data) {
    logger.error(`❌ Invalid or expired OAuth state: ${state}`);
    throw new Error("Invalid or expired state parameter.");
  }

  const expiry = new Date(data.expiry);
  if (new Date() > expiry) {
    logger.error(`❌ State expired: Generated: ${data.expiry} Current: ${new Date()}`);
    throw new Error("State token expired.");
  }

  return data;
};

module.exports = { validateState };
